### PR TITLE
Latent PARC

### LINKED ISSUE #55: porting latent parc

### PR DESCRIPTION
**Pull request by chengxinlun**
Integrating Latent PARC code with PARCtorch repo.

### CODE REVIEW

File: PARCtorch/autoencoder/autoencoder.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        super().__init__()
+        self.encoder = encoder
+        self.decoder = decoder
+
+    def forward(self, x):
+        encoded = self.encoder(x)
+        decoded = self.decoder(encoded)
+        return decoded
+    
+class Autoencoder_separate(nn.Module):
```
Comment (chengxinlun) [inline PARCtorch/autoencoder/autoencoder.py:136]: Please change class name according to camel case naming convention.

File: PARCtorch/autoencoder/train.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        return self.network(x)
+
+    def encode(self, x):
+        return self.network.encoder(x)
+
+    def decode(self, x):
+        return self.network.decoder(x)
+    
+
+def train_autoencoder(model, optimizer, loss_function, train_loader, val_loader,
```
Comment (chengxinlun) [inline PARCtorch/autoencoder/train.py:27]: Please add functionalities for resuming training from checkpoint.

File: PARCtorch/autoencoder/utils.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    Returns:
+        Noisy images.
+    """
+    noise = torch.rand_like(images) * (max_val - min_val) + min_val
+    noisy_images = images + noise
+    return torch.clamp(noisy_images, 0.0, 1.0)  # Keep pixel values in [0, 1]
+
+class LpLoss(torch.nn.Module):
+    def __init__(self, p=10):
```
[the remark this answers is not in the document]
Comment (chengxinlun) [reply PARCtorch/autoencoder/utils.py:35]: Also split the loss function part off, and put it under the module ```PARCtorch.loss```

File: PARCtorch/autoencoder/train.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            scheduler.step(avg_val_loss)
+
+    # Save model and logs
+    save_model(model, save_path, weights_name, epochs)
+    save_log(log_dict, save_path, weights_name, epochs)
+
+    return log_dict
+
+
+def train_individual_autoencoder(model, optimizer, loss_function, train_loader, val_loader,
```
Comment (chengxinlun) [inline PARCtorch/autoencoder/train.py:95]: Same as before. Please add functionalities for resuming training from checkpoint.

File: PARCtorch/autoencoder/train.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
+import torch
+import numpy as np
+from tqdm import tqdm
+import torch.nn.functional as F
```
Comment (chengxinlun) [inline PARCtorch/autoencoder/train.py:4]: Please remove unused import.

File: PARCtorch/autoencoder/autoencoder.py (hunk 1 above)
Comment (zgrayblue) [reply PARCtorch/autoencoder/autoencoder.py:136]: Done

File: PARCtorch/autoencoder/train.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        self.weights_name = weights_name
+
+    def autoencode(self, x):
+        return self.network(x)
+
+    def encode(self, x):
+        return self.network.encoder(x)
+
+    def decode(self, x):
+        return self.network.decoder(x)
```
[the remark this answers is not in the document]
Comment (zgrayblue) [reply PARCtorch/autoencoder/train.py:24]: Done

File: PARCtorch/autoencoder/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zgrayblue) [reply PARCtorch/autoencoder/utils.py:35]: Will do. Quick clarification: I have checked all the files and seems like a loss file doesn't exist yet. Should I create one?

File: PARCtorch/autoencoder/train.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
+import torch
+import numpy as np
+from tqdm import tqdm
+import torch.nn.functional as F
+
+from autoencoder import *
+from utils import *
```
[the remark this answers is not in the document]
Comment (zgrayblue) [reply PARCtorch/autoencoder/train.py:7]: Done

File: PARCtorch/autoencoder/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chengxinlun) [reply PARCtorch/autoencoder/utils.py]: > Will do. Quick clarification: I have checked all the files and seems like a loss file doesn't exist yet. Should I create one?

Yes, please create one.

[the remark this answers is not in the document]
Comment (chengxinlun) [reply PARCtorch/autoencoder/utils.py]: > Also I implemented the reduction input but found this online about the other two arguments: The size_average and reduce parameters are legacy arguments in PyTorch loss functions and are deprecated in favor of a single reduction argument.
> 
> Thoughts?

If they are deprecated, then just having ```reduction``` should be enough.
